File: src/main/java/frc/robot/util/interfaces/IMercMotorController.java
package frc.robot.util.interfaces;

import com.ctre.phoenix.motorcontrol.NeutralMode;

import frc.robot.util.PIDGain;

import com.ctre.phoenix.ParamEnum;
import com.ctre.phoenix.motorcontrol.FeedbackDevice;

public interface IMercMotorController {
    /**
     * Sets the relative speed of the controller. 
     * 
     * @param speed speed from -1 to 1
     */
    public void setSpeed(double x);

    /**
     * Sets relative position of controller
     * 
     * @param position the final position in ticks
     */
    public void setPosition(double x);

    /**
     * Want the speed of the controller?
     * 
     * @return the speed of the controller
     */
    public double getSpeed();

    /**
     * Sets a motor controller inverted.
     * 
     * @param invert Whether to invert or not
     */
    public void setInverted(boolean invert);

    /**
     * Follows the motor controller that is passed in. Will not do anything if it has to follow anything else.
     * 
     * @param leader the controller to follow
     */
    public void follow(IMercMotorController leader);

    /**
     * Want the CAN ID of this controller?
     * 
     * @return the CAN ID of this controller
     */
    public int getPort();

    /**
     * STOP! You have been caught singing despacito 9: yodeling kid edition
     */
    public void stop();

    /**
     * Want the encoder position of the controller?
     * NOTE: Victors can't connect to another sensor like an encoder, so it will return 0 for MercVictorSPX.
     * 
     * @return the encoder position of the controller
     */
    public double getEncTicks();

    /**
     * Want the encoder velocity of the controller?
     * NOTE: Victors can't connect to another sensor like an encoder, so it will return 0 for MercVictorSPX.
     * 
     * @return the encoder velocity of the controller
     */
    public double getEncVelo();

    /**
     * Want to reset the encoder?
     */
    public void resetEncoder();

    /**
     * Want the error from setpoint?
     * NOTE: Not to be used with Victors (no connected encoder)
     * 
     * @return the error from setpoint in ticks
     */
    public double getClosedLoopError();

    /**
     * Configure PID for the motor controller
     * 
     * @param slot slot to put it in
     * @param gains i got mad gains
     */
    public void configPID(int slot, PIDGain gains);

    /**
     * Configure the min and max voltage.
     * NOTE: for sparks, this is only for closed loop control. For talons, it is for both.
     * 
     * @param nominalOutput min voltage
     * @param peakOutput max voltage
     */
    public void configVoltage(double nominalOutput, double peakOutput);

    /**
     * Set the neutral mode - brake or coast
     * 
     * @param nm the NeutralMode to set it to
     */
    public void setNeutralMode(NeutralMode neutralMode);

    /**
     * 
     */
    public void setSensorPhase(boolean bool);

    /**
     * 
     */
    public void configAllowableClosedLoopError(int slotIdx, int allowableCloseLoopError, int timeoutMs);

    public void configSelectedFeedbackSensor(FeedbackDevice FeedbackDevice, int pidIdx, int timeoutMs);

    public void configSetParameter(ParamEnum param, double value, int subValue, int ordinal, int timeoutMs);

    public boolean isLimitSwitchClosed();
}

File: src/main/kotlin/ktfrc/robot/util/motorControllers/IMercMotorController.kt
package ktfrc.robot.util.motorControllers

import com.ctre.phoenix.ParamEnum
import com.ctre.phoenix.motorcontrol.*
import ktfrc.robot.util.PIDGain

interface IMercMotorController {

    //region Motor Control

    fun setPosition(position: Double)

    fun getEncTicks(): Double

    fun getEncVelocity(): Double

    fun setSpeed(speed: Double)

    fun getSpeed(): Double

    fun set(controlMode: ControlMode, demand0: Double, demand1Type: DemandType = DemandType.Neutral, demand1: Double = 0.0) //enums

    fun stop()

    //endregion

    //region Basic Motor Config

    fun configVoltage(nominalOutput: Double, peakOutput: Double)

    fun setInverted(invert: Boolean)

    fun setNeutralMode(neutralMode: NeutralMode) //Should be an enum (brake/coast)

    fun follow(leader: IMercMotorController, followerType: FollowerType = FollowerType.PercentOutput) //Make enum for follower type

    fun getPort(): Int

    fun configFactoryReset()

    //endregion

    //region Encoders/Limit Switches

    fun resetEncoder()

    fun setSensorPhase(phase: Boolean)

    fun isLimitSwitchClosed(polarity: LimitSwitchPolarity): Boolean

    //endregion

    //region Soft limit switches

    fun setForwardSoftLimit(limitTicks: Int)

    fun enableForwardSoftLimit()

    fun disableForwardSoftLimit()

    fun setReverseSoftLimit(limitTicks: Int)

    fun enableReverseSoftLimit()

    fun disableReverseSoftLimit()

    //endregion

    //region Closed Loop / PID

    fun configPID(slot: Int, gain: PIDGain)

    fun getClosedLoopError(slotIdx: Int = 0): Double

    fun configAllowableClosedLoopError(slotIdx: Int, allowableClosedLoopError: Int)

    fun configClosedLoopPeakOutput(slotIdx: Int, peakOutput: Double)

    fun configClosedLoopPeriod(slotIdx: Int, closedLoopTimeMs: Int)

    fun configAuxPIDPolarity(invert: Boolean)

    fun configSelectedFeedbackSensor(feedbackDevice: FeedbackDevice, pidIdx: Int) //enum

    fun configSetParameter(param: ParamEnum, value: Double, subValue: Int = 0, ordinal: Int = 0) //enum

    fun configSensorTerm(term: SensorTerm, feedbackDevice: FeedbackDevice)

    fun configRemoteFeedbackFilter(deviceID: Int, source: RemoteSensorSource, remoteSlotIdx: Int)

    fun configSelectedFeedbackCoefficient(feedbackScale: Double, pidIdx: Int)

    fun setStatusFramePeriod(statusFrame: StatusFrame, statusMs: Int)

    fun selectProfileSlot(slotIdx: Int, pidIdx: Int)

    //endregion

    //region Motion Magic

    fun configMotionAcceleration(sensorUnitsPer100msPerSec: Int)

    fun configMotionCruiseVelocity(sensorUnitsPer100ms: Int)

    //endregion
}

enum class LimitSwitchPolarity {
    FORWARD,
    REVERSE
}